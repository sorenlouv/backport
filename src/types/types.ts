import { AxiosError, AxiosResponse } from 'axios';

/*
 * Github
 */

export interface GithubQuery {
  access_token: string;
  per_page: number;
  author?: string;
}

export interface GithubIssue {
  html_url: string;
  number: number;
}

export interface GithubCommit {
  commit: {
    message: string;
  };
  sha: string;
}

export interface GithubSearch<T> {
  items: T[];
}

export interface GithubPullRequestPayload {
  title: string;
  head: string;
  base: string;
  body?: string;
  maintainer_can_modify?: boolean;
}

// TODO: Make PR to DefinitelyTypes to make AxiosError a generic that takes the error response as T
export interface GithubApiError extends AxiosError {
  response?: AxiosResponse<{
    message: string;
    errors?: Array<{}>;
    documentation_url: string;
  }>;
}

/*
 * Config
 */

interface ConfigOptions {
  all?: boolean;
  multiple?: boolean;
  multipleCommits?: boolean;
  multipleBranches?: boolean;
}

export interface ProjectConfigRaw extends ConfigOptions {
  upstream: string;
  branches: Array<string | BranchChoice>;
  labels?: string[];
}

export interface ProjectConfig extends ConfigOptions {
  upstream: string;
  branchChoices: BranchChoice[];
  labels?: string[];
}

export interface GlobalConfig extends ConfigOptions {
  username: string;
  accessToken: string;
}

export type CombinedConfig = Partial<GlobalConfig> & Partial<ProjectConfig>;

/*
 * Options
 */

export interface BackportOptions {
  accessToken: string;
  all: boolean;
  labels: string[];
  multiple: boolean;
  multipleBranches: boolean;
  multipleCommits: boolean;
  upstream: string;
  username: string;
  branchChoices?: BranchChoice[];
  branches?: string[];
  sha?: string;
}

/*
 * PullRequest
 */

export interface PullRequest {
  html_url: string;
  number: number;
}

/*
 * Commit
 */

export interface Commit {
  sha: string;
  message: string;
  pullRequest?: number;
}

/*
 * Branch
 */

export interface BranchChoice {
  name: string;
  checked?: boolean;
}
