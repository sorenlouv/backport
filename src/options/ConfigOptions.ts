import { Logger } from '../services/logger';

export interface BranchChoice {
  name: string;
  checked?: boolean;
}
export type BranchChoiceRaw = string | BranchChoice;

type AutoFixConflictsHandler = ({
  files,
  directory,
  logger,
  targetBranch,
}: {
  files: string[];
  directory: string;
  logger: Logger;
  targetBranch: string;
}) => boolean | Promise<boolean>;

export type ConfigOptions = Partial<{
  accessToken: string;
  all: boolean;
  assignees: string[];
  author: string;
  autoAssign: boolean;
  autoFixConflicts: AutoFixConflictsHandler;
  branchLabelMapping: Record<string, string>;
  ci: boolean;
  editor: string;
  forceLocalConfig: boolean;
  fork: boolean;
  gitHostname: string;
  githubApiBaseUrlV3: string;
  githubApiBaseUrlV4: string;
  maxNumber: number;
  multiple: boolean;
  multipleBranches: boolean;
  multipleCommits: boolean;
  noVerify: boolean;
  path: string;
  prDescription: string;
  prFilter: string;
  prTitle: string;
  pullNumber: number;
  resetAuthor: boolean;
  sha: string;
  sourceBranch: string;
  sourcePRLabels: string[];
  targetBranchChoices: BranchChoiceRaw[];
  targetBranches: string[];
  targetPRLabels: string[];
  upstream: string;
  username: string;
  verbose: boolean;

  // deprecated: renamed to `targetBranchChoices`
  branches: BranchChoiceRaw[];

  // deprecated: renamed to `targetPRLabels`
  labels: string[];
}>;
