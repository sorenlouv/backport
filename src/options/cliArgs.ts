import yargs from 'yargs';
import { OptionsFromConfigFiles } from './config/config';

export type OptionsFromCliArgs = ReturnType<typeof getOptionsFromCliArgs>;
export function getOptionsFromCliArgs(
  configOptions: OptionsFromConfigFiles,
  argv: readonly string[]
) {
  const cliArgs = yargs(argv)
    .parserConfiguration({
      'strip-dashed': true,
      'strip-aliased': true,
      'boolean-negation': false,
    })
    .usage('$0 [args]')
    .wrap(Math.max(100, Math.min(120, yargs.terminalWidth())))
    .option('accessToken', {
      alias: 'accesstoken',
      description: 'Github access token',
      type: 'string',
    })
    .option('all', {
      default: configOptions.all,
      description: 'List all commits',
      alias: 'a',
      type: 'boolean',
    })
    .option('author', {
      default: configOptions.author,
      description: 'Show commits by specific author',
      type: 'string',
    })
    .option('maxNumber', {
      default: configOptions.maxNumber,
      description: 'Number of commits to choose from',
      alias: ['number', 'n'],
      type: 'number',
    })
    .option('dryRun', {
      default: false,
      description: 'Perform backport without pushing to Github',
      type: 'boolean',
    })
    .option('editor', {
      default: configOptions.editor,
      description: 'Editor to be opened during conflict resolution',
      type: 'string',
    })
    .option('fork', {
      default: configOptions.fork,
      description: 'Create backports in fork or origin repo',
      type: 'boolean',
    })
    .option('gitHostname', {
      hidden: true,
      default: configOptions.gitHostname,
      description: 'Hostname for Github',
      type: 'string',
    })
    .option('githubApiBaseUrlV3', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV3,
      description: `Base url for Github's REST (v3) API`,
      type: 'string',
    })
    .option('githubApiBaseUrlV4', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV4,
      description: `Base url for Github's GraphQL (v4) API`,
      type: 'string',
    })
    .option('mainline', {
      description:
        'Parent id of merge commit. Defaults to 1 when supplied without arguments',
      type: 'number',
      coerce: (mainline) => {
        // `--mainline` (default to 1 when no parent is given)
        if (mainline === undefined) {
          return 1;
        }

        // use specified mainline parent
        if (Number.isInteger(mainline)) {
          return mainline as number;
        }

        // Invalid value provided
        throw new Error(`--mainline must be an integer. Received: ${mainline}`);
      },
    })
    .option('multiple', {
      default: configOptions.multiple,
      description: 'Select multiple branches/commits',
      type: 'boolean',
    })
    .option('multipleCommits', {
      default: configOptions.multipleCommits,
      description: 'Backport multiple commits',
      type: 'boolean',
    })
    .option('multipleBranches', {
      default: configOptions.multipleBranches,
      description: 'Backport to multiple branches',
      type: 'boolean',
    })
    .option('noVerify', {
      default: configOptions.noVerify,
      description: 'Bypasses the pre-commit and commit-msg hooks',
      type: 'boolean',
    })
    .option('path', {
      default: configOptions.path,
      description: 'Only list commits touching files under the specified path',
      alias: 'p',
      type: 'string',
    })
    .option('prTitle', {
      default: configOptions.prTitle,
      description: 'Title of pull request',
      alias: 'title',
      type: 'string',
    })
    .option('prDescription', {
      default: configOptions.prDescription,
      description: 'Description to be added to pull request',
      alias: 'description',
      type: 'string',
    })
    .option('pullNumber', {
      conflicts: ['sha', 'sourcePRsFilter'],
      description: 'Pull request to backport',
      type: 'number',
      alias: 'pr',
    })
    .option('resetAuthor', {
      default: false,
      description: 'Set yourself as commit author',
      type: 'boolean',
    })
    .option('sha', {
      conflicts: ['pullNumber', 'sourcePRsFilter'],
      description: 'Commit sha to backport',
      type: 'string',
      alias: 'commit',
    })
    .option('sourcePRLabels', {
      default: configOptions.sourcePRLabels,
      description: 'Add labels to the source (original) PR',
      type: 'array',
      alias: 'sourcePRLabel',
    })
    .option('sourcePRsFilter', {
      conflicts: ['pullNumber', 'sha'],
      // default: configOptions.githubApiBaseUrlV4,
      description: `Filter source pull requests by a query`,
      alias: 'pr-filter',
      type: 'string',
    })
    .option('sourceBranch', {
      default: configOptions.sourceBranch,
      description: `List commits to backport from another branch than master`,
      type: 'string',
    })
    .option('verify', {
      description: `Opposite of no-verify`,
      type: 'boolean',
    })
    .option('targetBranches', {
      default: [] as string[],
      description: 'Branch(es) to backport to',
      type: 'array',
      alias: ['targetBranch', 'branch', 'b'],
      string: true, // ensure `6.0` is not coerced to `6`
    })
    .option('targetPRLabels', {
      default: configOptions.targetPRLabels,
      description: 'Add labels to the target (backport) PR',
      alias: ['labels', 'label', 'l'],
      type: 'array',
    })
    .option('upstream', {
      default: configOptions.upstream,
      description: 'Name of repository',
      alias: 'up',
      type: 'string',
    })
    .option('username', {
      default: configOptions.username,
      description: 'Github username',
      type: 'string',
    })
    .option('verbose', {
      default: false,
      description: 'Show additional debug information',
      type: 'boolean',
    })
    .alias('version', 'v')
    .alias('version', 'V')
    .help()
    .epilogue(
      'For bugs, feature requests or questions: https://github.com/sqren/backport/issues\nOr contact me directly: https://twitter.com/sorenlouv'
    ).argv;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  const { $0, _, verify, ...rest } = cliArgs;

  return {
    ...rest,
    accessToken: cliArgs.accessToken || configOptions.accessToken, // accessToken should not be displayed in yargs help menu
    branchLabelMapping: configOptions.branchLabelMapping, // not available as cli argument
    multipleBranches: cliArgs.multipleBranches || cliArgs.multiple,
    multipleCommits: cliArgs.multipleCommits || cliArgs.multiple,
    noVerify: verify ?? rest.noVerify, // `verify` is a cli-only flag to flip the default of `no-verify`
    targetBranchChoices: configOptions.targetBranchChoices, // not available as cli argument
  };
}
