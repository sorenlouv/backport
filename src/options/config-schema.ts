import { z } from 'zod';
import type { TargetBranchChoiceOrString } from './config-options';

// Target branch choice schema
const targetBranchChoiceSchema: z.ZodType<TargetBranchChoiceOrString> = z.union(
  [
    z.string(),
    z.object({
      name: z.string(),
      checked: z.boolean().optional(),
    }),
  ],
);

// Base config schema WITHOUT defaults - used for parsing
// Defaults are applied later in the merging process
const baseConfigSchema = z.object({
  accessToken: z.string().optional(),
  assignees: z.array(z.string()).optional(),
  author: z.union([z.string(), z.null()]).optional(),
  autoAssign: z.boolean().optional(),
  autoMerge: z.boolean().optional(),
  autoMergeMethod: z.enum(['merge', 'rebase', 'squash']).optional(),
  backportBinary: z.string().optional(),
  backportBranchName: z.string().optional(),
  cherrypickRef: z.boolean().optional(),
  commitConflicts: z.boolean().optional(),
  commitPaths: z.array(z.string()).optional(),
  copySourcePRLabels: z.boolean().optional(),
  copySourcePRReviewers: z.boolean().optional(),
  details: z.boolean().optional(),
  dir: z.string().optional(),
  draft: z.boolean().optional(),
  dryRun: z.boolean().optional(),
  editor: z.string().optional(),
  fork: z.boolean().optional(),
  gitAuthorEmail: z.string().optional(),
  gitAuthorName: z.string().optional(),
  gitHostname: z.string().optional(),
  githubActionRunId: z.number().optional(),
  githubApiBaseUrlV3: z.string().optional(),
  githubApiBaseUrlV4: z.string().optional(),
  interactive: z.boolean().optional(),
  logFilePath: z.string().optional(),
  ls: z.boolean().optional(),
  maxNumber: z.number().optional(),
  mainline: z.number().optional(),
  multiple: z.boolean().optional(),
  multipleBranches: z.boolean().optional(),
  multipleCommits: z.boolean().optional(),
  noVerify: z.boolean().optional(),
  prDescription: z.string().optional(),
  prFilter: z.string().optional(),
  projectConfigFile: z.string().optional(),
  prTitle: z.string().optional(),
  publishStatusCommentOnAbort: z.boolean().optional(),
  publishStatusCommentOnFailure: z.boolean().optional(),
  publishStatusCommentOnSuccess: z.boolean().optional(),
  pullNumber: z.union([z.number(), z.array(z.number())]).optional(),
  repoForkOwner: z.string().optional(),
  repoName: z.string().optional(),
  repoOwner: z.string().optional(),
  resetAuthor: z.boolean().optional(),
  reviewers: z.array(z.string()).optional(),
  sha: z.union([z.string(), z.array(z.string())]).optional(),
  signoff: z.boolean().optional(),
  skipRemoteConfig: z.boolean().optional(),
  noUnmergedBackportsHelp: z.boolean().optional(),
  sourceBranch: z.string().optional(),
  sourcePRLabels: z.array(z.string()).optional(),
  targetBranchChoices: z.array(targetBranchChoiceSchema).optional(),
  targetBranches: z.array(z.string()).optional(),
  targetPRLabels: z.array(z.string()).optional(),
  telemetry: z.boolean().optional(),
  cwd: z.string().optional(),
  dateSince: z.string().nullable().optional(),
  dateUntil: z.string().nullable().optional(),
  globalConfigFile: z.string().optional(),

  // Deprecated options (for backward compatibility)
  upstream: z.string().optional(),
  branches: z.array(targetBranchChoiceSchema).optional(),
  labels: z.array(z.string()).optional(),
  addOriginalReviewers: z.boolean().optional(),

  // Config file only options
  branchLabelMapping: z.record(z.string(), z.string()).optional(),

  // yargs-only options (for parsing, not for validation)
  help: z.boolean().optional(),
  version: z.boolean().optional(),
  v: z.boolean().optional(),
});

// Schema for partial config (from files/cli) - without defaults applied
// This is used for parsing CLI args and config files where we don't want defaults
// Defaults are applied later via manual merge in options.ts
export const partialConfigSchema = baseConfigSchema.loose();
