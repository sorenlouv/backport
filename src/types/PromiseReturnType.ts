export type PromiseReturnType<Func> = Func extends (
  ...args: any[]
) => Promise<infer Value>
  ? Value
  : Func;
