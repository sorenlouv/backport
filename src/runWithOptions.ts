import { BackportOptions } from './options/options';
import { getBranches } from './ui/getBranches';
import { getCommits } from './ui/getCommits';
import { maybeSetupRepo } from './ui/maybeSetupRepo';
import { logger } from './services/logger';
import { cherrypickAndCreatePullRequest } from './ui/cherrypickAndCreatePullRequest';
import { sequentially } from './services/sequentially';
import { HandledError } from './services/HandledError';
import { addLabelsToPullRequest } from './services/github/addLabelsToPullRequest';
import { withSpinner } from './ui/withSpinner';

export async function runWithOptions(options: BackportOptions) {
  const commits = await getCommits(options);
  const branches = await getBranches(options);

  await maybeSetupRepo(options);

  let backportSucceeded = false; // minimum 1 backport PR was successfully created
  await sequentially(branches, async baseBranch => {
    logger.info(`Backporting ${JSON.stringify(commits)} to ${baseBranch}`);
    try {
      await cherrypickAndCreatePullRequest({ options, commits, baseBranch });
      backportSucceeded = true;
    } catch (e) {
      if (e instanceof HandledError) {
        console.error(e.message);
      } else {
        console.error(e);
        throw e;
      }
    }
  });

  if (backportSucceeded && options.backportCreatedLabels.length > 0) {
    await Promise.all(
      commits.map(async ({ pullNumber }) => {
        if (pullNumber) {
          return withSpinner(
            { text: `Adding labels to #${pullNumber}` },
            () => {
              return addLabelsToPullRequest(
                options,
                pullNumber,
                options.backportCreatedLabels
              );
            }
          );
        }
      })
    );
  }
}
