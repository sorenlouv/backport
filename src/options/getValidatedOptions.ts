import isEmpty from 'lodash.isempty';
import { HandledError } from '../services/HandledError';
import { getGlobalConfigPath } from '../services/env';
import { OptionsFromCliArgs } from './cliArgs';

const GLOBAL_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson';

const PROJECT_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson';

export type ValidatedOptions = ReturnType<typeof getValidatedOptions>;
export function getValidatedOptions({
  // omit `upstream` from options
  upstream,
  ...options
}: OptionsFromCliArgs) {
  // ensure accessToken and username are given
  if (!options.accessToken || !options.username) {
    const globalConfigPath = getGlobalConfigPath();
    throw new HandledError(
      `Please update your config file: ${globalConfigPath}.\nIt must contain a valid "username" and "accessToken".\n\nRead more: ${GLOBAL_CONFIG_DOCS_LINK}`
    );
  }

  // ensure targetBranches or targetBranchChoices are given
  if (isEmpty(options.targetBranches) && isEmpty(options.targetBranchChoices)) {
    throw new HandledError(
      `You must specify a target branch\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--branch 6.1"`
    );
  }

  // ensure a valid upstream is given
  const [repoOwner, repoName] = (upstream ?? '').split('/');
  if (!repoOwner || !repoName) {
    throw new HandledError(
      `You must specify a valid Github repository\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--upstream elastic/kibana"`
    );
  }

  return {
    ...options,

    // no longer optional
    accessToken: options.accessToken,
    username: options.username,

    // split upstream
    repoName,
    repoOwner,

    // define author
    author: options.author || options.username,
    all: options.author ? false : options.all,
  };
}
