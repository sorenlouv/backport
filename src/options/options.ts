import { fetchDefaultRepoBranchAndPerformStartupChecks } from '../services/github/v4/fetchDefaultRepoBranchAndPerformStartupChecks';
import { PromiseReturnType } from '../types/PromiseReturnType';
import { setLogLevel, setRedactedAccessToken } from './../services/logger';
import { ConfigOptions } from './ConfigOptions';
import { getOptionsFromCliArgs } from './cliArgs';
import { getOptionsFromConfigFiles } from './config/config';
import { getValidatedOptions } from './getValidatedOptions';

export type BackportOptions = Readonly<PromiseReturnType<typeof getOptions>>;
export async function getOptions(
  argv: string[],
  optionsFromModule?: ConfigOptions
) {
  const optionsFromConfig = await getOptionsFromConfigFiles(
    optionsFromModule?.ci
  );
  const optionsFromCli = getOptionsFromCliArgs(
    { ...optionsFromConfig, ...optionsFromModule },
    argv
  );
  setRedactedAccessToken(optionsFromCli.accessToken);

  // set log level when all config options have been taken into account
  setLogLevel({ verbose: optionsFromCli.verbose });

  // TODO: move `getValidatedOptions` to `getOptionsFromCliArgs`
  const validatedOptions = getValidatedOptions(optionsFromCli);

  // TODO: make `username` optional by defaulting to `currentUsername`
  const { defaultBranch } = await fetchDefaultRepoBranchAndPerformStartupChecks(
    validatedOptions
  );

  return {
    ...validatedOptions,

    // use the default branch as source branch (normally master) unless an explicit `sourceBranch` has been given
    sourceBranch: validatedOptions.sourceBranch
      ? validatedOptions.sourceBranch
      : defaultBranch,
  };
}
