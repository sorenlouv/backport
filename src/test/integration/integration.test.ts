import { once } from 'lodash';
import nock from 'nock';
import { getOptions } from '../../options/options';
import { runWithOptions } from '../../runWithOptions';
import { PromiseReturnType } from '../../types/PromiseReturnType';
import { createSpies } from './createSpies';
import { REMOTE_ORIGIN_REPO_PATH, REMOTE_FORK_REPO_PATH } from './envConstants';
import {
  getBranches,
  getLatestCommit,
  deleteAndSetupEnvironment,
} from './helpers';

jest.unmock('make-dir');
jest.unmock('del');
jest.unmock('../../services/child-process-promisified');

describe('integration', () => {
  afterAll(() => {
    nock.cleanAll();
  });

  describe('when a single commit is backported', () => {
    let res: PromiseReturnType<typeof runWithOptions>;
    let spies: ReturnType<typeof createSpies>;

    beforeEach(
      once(async () => {
        jest.clearAllMocks();
        spies = createSpies({ commitCount: 1 });

        await deleteAndSetupEnvironment();

        const options = await getOptions([], {
          // use localhost to avoid CORS issues
          githubApiBaseUrlV4: 'http://localhost/graphql',
        });
        res = await runWithOptions(options);
      })
    );

    it('returns pull request', () => {
      expect(res).toEqual([
        { pullRequestUrl: 'myHtmlUrl', success: true, targetBranch: '6.0' },
      ]);
    });

    it('sends the correct http body when creating pull request', () => {
      expect(spies.createPullRequestCalls).toEqual([
        {
          base: '6.0',
          body: 'Backports the following commits to 6.0:\n - Add witch (#85)',
          head: 'sqren:backport/6.0/pr-85',
          title: '[6.0] Add witch (#85)',
        },
      ]);
    });

    it('should make correct API requests', () => {
      expect(spies.getGithubConfigOptionsCalls).toMatchSnapshot();
      expect(spies.authorIdCalls).toMatchSnapshot();
      expect(spies.commitsByAuthorCalls).toMatchSnapshot();
    });

    it('should not create new branches in origin (backport-org/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_ORIGIN_REPO_PATH);
      expect(branches).toEqual(['6.0', '* master']);
    });

    it('should create branch in the fork (sqren/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_FORK_REPO_PATH);
      expect(branches).toEqual(['6.0', 'backport/6.0/pr-85', '* master']);
    });

    it('should have cherry picked the correct commit', async () => {
      const commit = await getLatestCommit({
        branch: 'backport/6.0/pr-85',
        commitCount: 1,
        cwd: REMOTE_FORK_REPO_PATH,
      });
      expect(commit).toMatchInlineSnapshot(`
      " romeo-and-juliet.txt | 2 +-
       1 file changed, 1 insertion(+), 1 deletion(-)

      diff --git a/romeo-and-juliet.txt b/romeo-and-juliet.txt
      index 87f1ac7..51e1e4b 100644
      --- a/romeo-and-juliet.txt
      +++ b/romeo-and-juliet.txt
      @@ -158 +158 @@ Thereto prick'd on by a most emulate pride,
      -Dared to the combat; in which our valiant Hamlet--
      +Dared to the combat; in üßô‚Äç‚ôÄÔ∏è our valiant Hamlet--
      "
    `);
    });
  });

  describe('when two commits are backported', () => {
    let spies: ReturnType<typeof createSpies>;
    let res: PromiseReturnType<typeof runWithOptions>;

    beforeEach(
      once(async () => {
        jest.clearAllMocks();
        spies = createSpies({ commitCount: 2 });

        await deleteAndSetupEnvironment();

        const options = await getOptions([], {
          // use localhost to avoid CORS issues
          githubApiBaseUrlV4: 'http://localhost/graphql',
        });
        res = await runWithOptions(options);
      })
    );

    it('sends the correct http body when creating pull request', () => {
      expect(spies.createPullRequestCalls).toEqual([
        {
          title: '[6.0] Add witch (#85) | Add üëª (2e63475c)',
          head: 'sqren:backport/6.0/pr-85_commit-2e63475c',
          base: '6.0',
          body: 'Backports the following commits to 6.0:\n - Add witch (#85)\n - Add üëª (2e63475c)',
        },
      ]);
    });

    it('returns pull request', () => {
      expect(res).toEqual([
        { pullRequestUrl: 'myHtmlUrl', success: true, targetBranch: '6.0' },
      ]);
    });

    it('should make correct API requests', () => {
      expect(spies.getGithubConfigOptionsCalls).toMatchSnapshot();
      expect(spies.authorIdCalls).toMatchSnapshot();
      expect(spies.commitsByAuthorCalls).toMatchSnapshot();
    });

    it('should not create new branches in origin (backport-org/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_ORIGIN_REPO_PATH);
      expect(branches).toEqual(['6.0', '* master']);
    });

    it('should create branch in the fork (sqren/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_FORK_REPO_PATH);
      expect(branches).toEqual([
        '6.0',
        'backport/6.0/pr-85_commit-2e63475c',
        '* master',
      ]);
    });

    it('should have cherry picked the correct commit', async () => {
      const commit = await getLatestCommit({
        branch: 'backport/6.0/pr-85_commit-2e63475c',
        commitCount: 2,
        cwd: REMOTE_FORK_REPO_PATH,
      });
      expect(commit).toMatchInlineSnapshot(`
            " romeo-and-juliet.txt | 2 +-
             1 file changed, 1 insertion(+), 1 deletion(-)

            diff --git a/romeo-and-juliet.txt b/romeo-and-juliet.txt
            index 51e1e4b..4814bb0 100644
            --- a/romeo-and-juliet.txt
            +++ b/romeo-and-juliet.txt
            @@ -203 +203 @@ But soft, behold! lo, where it comes again!
            -Re-enter Ghost
            +Re-enter üëª
             romeo-and-juliet.txt | 2 +-
             1 file changed, 1 insertion(+), 1 deletion(-)

            diff --git a/romeo-and-juliet.txt b/romeo-and-juliet.txt
            index 87f1ac7..51e1e4b 100644
            --- a/romeo-and-juliet.txt
            +++ b/romeo-and-juliet.txt
            @@ -158 +158 @@ Thereto prick'd on by a most emulate pride,
            -Dared to the combat; in which our valiant Hamlet--
            +Dared to the combat; in üßô‚Äç‚ôÄÔ∏è our valiant Hamlet--
            "
        `);
    });
  });

  describe('when disabling fork mode', () => {
    let res: PromiseReturnType<typeof runWithOptions>;
    let spies: ReturnType<typeof createSpies>;

    beforeEach(
      once(async () => {
        jest.clearAllMocks();

        spies = createSpies({ commitCount: 1 });
        await deleteAndSetupEnvironment();

        const options = await getOptions(['--fork=false'], {
          // use localhost to avoid CORS issues
          githubApiBaseUrlV4: 'http://localhost/graphql',
        });
        res = await runWithOptions(options);
      })
    );

    it('sends the correct http body when creating pull request', () => {
      expect(spies.createPullRequestCalls).toEqual([
        {
          base: '6.0',
          body: 'Backports the following commits to 6.0:\n - Add witch (#85)',
          head: 'backport-org:backport/6.0/pr-85',
          title: '[6.0] Add witch (#85)',
        },
      ]);
    });

    it('returns pull request', () => {
      expect(res).toEqual([
        { pullRequestUrl: 'myHtmlUrl', success: true, targetBranch: '6.0' },
      ]);
    });

    it('should create new branches in origin (backport-org/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_ORIGIN_REPO_PATH);
      expect(branches).toEqual(['6.0', 'backport/6.0/pr-85', '* master']);
    });

    it('should NOT create branch in the fork (sqren/backport-demo)', async () => {
      const branches = await getBranches(REMOTE_FORK_REPO_PATH);
      expect(branches).toEqual(['6.0', '* master']);
    });

    it('should cherry pick the correct commit', async () => {
      const commit = await getLatestCommit({
        branch: 'backport/6.0/pr-85',
        commitCount: 1,
        cwd: REMOTE_ORIGIN_REPO_PATH,
      });
      expect(commit).toMatchInlineSnapshot(`
      " romeo-and-juliet.txt | 2 +-
       1 file changed, 1 insertion(+), 1 deletion(-)

      diff --git a/romeo-and-juliet.txt b/romeo-and-juliet.txt
      index 87f1ac7..51e1e4b 100644
      --- a/romeo-and-juliet.txt
      +++ b/romeo-and-juliet.txt
      @@ -158 +158 @@ Thereto prick'd on by a most emulate pride,
      -Dared to the combat; in which our valiant Hamlet--
      +Dared to the combat; in üßô‚Äç‚ôÄÔ∏è our valiant Hamlet--
      "
    `);
    });
  });
});
