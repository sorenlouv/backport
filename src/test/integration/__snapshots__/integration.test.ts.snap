// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when a single commit is backported should make correct API requests 1`] = `
Array [
  Object {
    "query": "
    query DefaultRepoBranch($repoOwner: String!, $repoName: String!) {
      viewer {
        login
      }
      repository(owner: $repoOwner, name: $repoName) {
        # check whether \\"backport\\" branch exists
        ref(qualifiedName: \\"refs/heads/backport\\") {
          name
        }

        # get default branch
        defaultBranchRef {
          name
        }
      }
    }
  ",
    "variables": Object {
      "repoName": "backport-demo",
      "repoOwner": "elastic",
    },
  },
]
`;

exports[`when a single commit is backported should make correct API requests 2`] = `
Array [
  Object {
    "query": "
    query AuthorId($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
]
`;

exports[`when a single commit is backported should make correct API requests 3`] = `
Array [
  Object {
    "query": "
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $historyPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $historyPath
              ) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          repository {
                            owner {
                              login
                            }
                            name
                          }
                          number
                          mergeCommit {
                            oid
                          }
                          labels(first: 50) {
                            nodes {
                              name
                            }
                          }
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      title
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
    "variables": Object {
      "authorId": "sqren_author_id",
      "historyPath": null,
      "maxNumber": 10,
      "repoName": "backport-demo",
      "repoOwner": "elastic",
      "sourceBranch": "master",
    },
  },
]
`;

exports[`when two commits are backported should make correct API requests 1`] = `
Array [
  Object {
    "query": "
    query DefaultRepoBranch($repoOwner: String!, $repoName: String!) {
      viewer {
        login
      }
      repository(owner: $repoOwner, name: $repoName) {
        # check whether \\"backport\\" branch exists
        ref(qualifiedName: \\"refs/heads/backport\\") {
          name
        }

        # get default branch
        defaultBranchRef {
          name
        }
      }
    }
  ",
    "variables": Object {
      "repoName": "backport-demo",
      "repoOwner": "elastic",
    },
  },
]
`;

exports[`when two commits are backported should make correct API requests 2`] = `
Array [
  Object {
    "query": "
    query AuthorId($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
]
`;

exports[`when two commits are backported should make correct API requests 3`] = `
Array [
  Object {
    "query": "
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $historyPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $historyPath
              ) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          repository {
                            owner {
                              login
                            }
                            name
                          }
                          number
                          mergeCommit {
                            oid
                          }
                          labels(first: 50) {
                            nodes {
                              name
                            }
                          }
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      title
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
    "variables": Object {
      "authorId": "sqren_author_id",
      "historyPath": null,
      "maxNumber": 10,
      "repoName": "backport-demo",
      "repoOwner": "elastic",
      "sourceBranch": "master",
    },
  },
]
`;
